obstore/python/obstore/auth/earthdata.py

```diff
@@ -6,157 +6,419 @@
+        finally:
+            await credential_provider.close()
+        ```
 
     [NASA Earthdata]: https://www.earthdata.nasa.gov/
+    [in-region direct S3 access]: https://archive.podaac.earthdata.nasa.gov/s3credentialsREADME
+
     """  # noqa: E501
 
     config: S3Config
 
     def __init__(
         self,
-        username: str,
-        password: str,
+        credentials_url: str,
+        *,
+        host: str | None = None,
+        auth: str | tuple[str, str] | None = None,
+        session: aiohttp.ClientSession | aiohttp_retry.RetryClient | None = None,
     ) -> None:
-        """Create a new NasaEarthdataAsyncCredentialProvider.
+        """Construct a new NasaEarthdataAsyncCredentialProvider.
 
-        Args:
-            username: Username to NASA Earthdata.
-            password: Password to NASA Earthdata.
+        This credential provider uses `aiohttp`, and will error if that cannot be
+        imported.
 
-        """
-        from aiohttp import BasicAuth, ClientSession
+        Refer to
+        [NasaEarthdataCredentialProvider][obstore.auth.earthdata.NasaEarthdataCredentialProvider.__init__]
+        for argument explanations.
 
-        # Pass region default
+        """
         self.config = {"region": "us-west-2"}
-        self.session = ClientSession(auth=BasicAuth(username, password))
+        self._session: aiohttp.ClientSession | aiohttp_retry.RetryClient | None = (
+            session or default_aiohttp_session()
+        )
+        # Avoid closing a user-supplied session (the user is responsible for that)
+        self._close = None if session else self._session.close
+        self._get_credentials_async = _make_get_credentials_async(
+            credentials_url,
+            host or _default_host(),
+            auth or _default_auth(),
+        )
 
     async def __call__(self) -> S3Credential:
         """Request updated credentials."""
-        async with self.session.get(CREDENTIALS_API, allow_redirects=True) as resp:
-            auth_url = resp.url
-        async with self.session.get(auth_url, allow_redirects=True) as auth_resp:
-            # Note: We parse the JSON manually instead of using `resp.json()` because
-            # the response mimetype is incorrectly set to text/html.
-            creds = json.loads(await auth_resp.text())
-        return {
-            "access_key_id": creds["accessKeyId"],
-            "secret_access_key": creds["secretAccessKey"],
-            "token": creds["sessionToken"],
-            "expires_at": datetime.fromisoformat(creds["expiration"]),
-        }
+        if self._session is None:
+            msg = "credential provider was closed"
+            raise ValueError(msg)
+
+        metadata = await self._get_credentials_async(self._session)
+
+        return _metadata_to_s3_credential(metadata)
 
     async def close(self) -> None:
-        """Close the underlying session.
+        """Close the underlying session, if it was not supplied by the user.
```

Comment:
Are there any external references we can consider here? I'm inclined to make `close` _always_ close the underlying session, and the user can decide not to call `close` if they want to keep using the underlying session. Seems like that would be more predictable.